<html>
<head>
<title>Mysql User Defined Functions for Bioinformatics</title>
<style>
body	{
	background-color:white;
	color:black;
	margin:10px;
	padding:10px;
	}

h1,h2,h3,h4	{
	font-family:Arial,Helvetica,Sans-Serif;
	margin:10px;
	padding:10px;
	}

h1	{
	background-color:blue;
	color:white;
	text-align:center;
	font-size:200%;
	}

h3	{
	margin:10px;
	padding:10px;
	font-size:200%;
	background-color:lightgreen;
	}

pre	{
	background-color:lightgray;
	font-size:150%;
	padding:10px;
	}
</style>
</head>
<body>
<h1>Mysql User Defined Functions for Bioinformatics</h1>
<h2>Introduction</h2>
<h3>Edit <code>make.properties</code></h3>
mysql.libdir is where the library should be saved For example<pre>/usr/lib</pre>On mysql <b>5.1.37</b> you can get it via:<pre>mysql_config --plugindir</pre>
data.dir should be a full path directory, readeable by mysql.
<h3>AppArmor</h3>
<pre> sudo /etc/init.d/apparmor stop</pre>
<h2>BioUDFs</h2>

<div>
<h3>faidx</h3>
<h4>About</h4><p><b>faidx</b> retrieves a subsequence of a larger FASTA sequence. The path of one or more fasta sequence must be defined in <code>make.properties</code> as <code>fasta.files</code>. For example:</p>
<pre>fasta.files=/data/ucsc/hg18/*.fa</pre>.
<h4>Build</h4><pre>make faidx</pre>
<h4>Create Function</h4>
<pre>mysql&gt; create function faidx returns string soname "faidx.so";
Query OK, 0 rows affected (0.00 sec)</pre>
<h4>Invoke Function</h4>
<pre>mysql&gt; select faidx("gi|10281952",2,20);
+---------------------------+
| faidx("gi|10281952",2,20) |
+---------------------------+
| CGACCCTTGCGTCCGTGA        | 
+---------------------------+
1 row in set (0.00 sec)

mysql&gt;select faidx("gi|10281952",2,200000);
+-------------------------------+
| faidx("gi|10281952",2,200000) |
+-------------------------------+
| NULL                          | 
+-------------------------------+
1 row in set (0.00 sec)

mysql&gt; select faidx("gi|10281952",100,200);
+------------------------------------------------------------------------------------------------------+
| faidx("gi|10281952",100,200)                                                                         |
+------------------------------------------------------------------------------------------------------+
| AGCCGCGACGTGATTAATGATATCACTGAGGCTAGGCAGGTTGTTGCAAATTTACGTGGACACCTGGGTGGCTCCGGCTTCAATGTGATAGCTTTGGCGA | 
+------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

</pre>
<h4>Drop Function</h4>
<pre>DROP FUNCTION faidx;
Query OK, 0 rows affected (0.00 sec)</pre>

</div>


<div>
<h3>taxon_name</h3>
<h4>About</h4><p>Get a label for a taxon in the <a href="http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/">NCBI taxonomy</a>.</p>
<h4>Create Function</h4>
<pre>create function taxon_name returns string soname "taxonudf.so";
Query OK, 0 rows affected (0.03 sec)
</pre>
<h4>Invoke Function</h4>
<pre>select taxon,<b>taxon_name</b>(taxon) from t1 where taxon_childof(taxon,taxon_id("Hominidae"));

+-------+-------------------------------+
| taxon | taxon_name(taxon)             |
+-------+-------------------------------+
|  9605 | Homo                          |
|  9606 | Homo sapiens                  |
| 63221 | Homo sapiens neanderthalensis |
+-------+-------------------------------+
3 rows in set (0.00 sec)
</pre>
<h4>Drop Function</h4>
<pre>DROP FUNCTION taxon_name;
Query OK, 0 rows affected (0.00 sec)</pre>
</div>


<div>
<h3>taxon_id</h3>
<h4>About</h4><p>Get the id of a taxon from a scientific name.</p>
<h4>Create Function</h4>
<pre>create function taxon_id returns integer soname "taxonudf.so";
Query OK, 0 rows affected (0.03 sec)
</pre>
<h4>Invoke Function</h4>
<pre>select taxon,taxon_name(taxon) from t1 where taxon_childof(taxon,<b>taxon_id</b>("Hominidae"));

+-------+-------------------------------+
| taxon | taxon_name(taxon)             |
+-------+-------------------------------+
|  9605 | Homo                          |
|  9606 | Homo sapiens                  |
| 63221 | Homo sapiens neanderthalensis |
+-------+-------------------------------+
3 rows in set (0.00 sec)


</pre>
<h4>Drop Function</h4>
<pre>DROP FUNCTION taxon_id;
Query OK, 0 rows affected (0.00 sec)</pre>
</div>


<div>
<h3>taxon_childof</h3>
<h4>About</h4><p>Get the common ancestor of two <a href="http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/">taxon records</a>.</p>
<h4>Create Function</h4>
<pre>create function taxon_childof returns integer soname "taxonudf.o";
Query OK, 0 rows affected (0.03 sec)
</pre>
<h4>Invoke Function</h4>
<pre>select taxon,taxon_name(taxon) from t1 where <b>taxonChildOf</b>(taxon,taxonId("Hominidae"));

+-------+-------------------------------+
| taxon | taxon_name(taxon)             |
+-------+-------------------------------+
|  9605 | Homo                          |
|  9606 | Homo sapiens                  |
| 63221 | Homo sapiens neanderthalensis |
+-------+-------------------------------+
3 rows in set (0.00 sec)


</pre>
<h4>Drop Function</h4>
<pre>DROP FUNCTION taxonName;
Query OK, 0 rows affected (0.00 sec)</pre>
</div>



<div>
<h3>Translate</h3>
<h4>About</h4><p>Translate a DNA to a protein using the standard genetic code.</p>
<h4>Create Function</h4>
<pre>CREATE FUNCTION translate RETURNS STRING SONAME 'translate.so';
Query OK, 0 rows affected (0.03 sec)
</pre>
<h4>Invoke Function</h4>
<pre>select translate("atggagaggcagaaacggaaggcggacatcgagaaagggctgca") as "roxan";
+----------------+
| roxan          |
+----------------+
| MERQKRKADIEKGL |
+----------------+
1 row in set (0.00 sec)

</pre>
<h4>Drop Function</h4>
<pre>DROP FUNCTION translate;
Query OK, 0 rows affected (0.00 sec)</pre>
</div>
<hr/>
<a href="http://plindenbaum.blogspot.com">Pierre Lindenbaum PhD.</a> 2010.
</body>
</html>
